### ISSUE TITLE
BoringSSL assembly + Visual Studio (probably) works now

### ISSUE THREAD

**Issue by davidben** [grpc/grpc#16376]
I noticed you all's documentation says that building with Visual Studio disables BoringSSL assembly, I'm guessing due to this bit here:
https://github.com/grpc/grpc/blob/2336c69d32d63ab87c2269c5a9eae2cba0c2cf0b/cmake/ssl.cmake#L20

FYI, as far as I know (I tested it with VS2017 when reviewing the workaround), this works fine these days, as long as your BoringSSL is up-to-date (June 29). An external contributor sent in a workaround recently:
https://boringssl-review.googlesource.com/c/boringssl/+/29504

It turns out it's actually a CMake bug, which we fixed upstream:
https://gitlab.kitware.com/cmake/cmake/merge_requests/2179

**Comment by sreecha** [comment]
@jtattermusch : Not sure if you are familiar with the context here (I am not). Assigning this to you for now :).

**Comment by jtattermusch** [comment]
@davidben  thanks for the heads up! This is good to know, I'll see if we can make setting OPENSSL_NO_ASM optional. The default will probably need to stay `OPENSSL_NO_ASM ON` though, because we are shipping the pre-built binaries in some of the packages (e.g. C# nugets) and making those pre-build binaries CPU-architecture dependent could break users.
(see https://github.com/grpc/grpc/blob/master/doc/ssl-performance.md for an overview).

**Comment by jtattermusch** [comment]
Note: it seems that the chromium-stable branch of boringssl (https://github.com/google/boringssl/tree/chromium-stable) doesn't contain https://boringssl-review.googlesource.com/c/boringssl/+/29504 yet, so the fix would need to be included in that branch before we can update our third_party/boringssl submodule to obtain the fix.

**Comment by jtattermusch** [comment]
@davidben  do you want to backport the fix to the chromium-stable branch?

**Comment by jtattermusch** [comment]
As an experiment, it tried disabling OPENSSL_NO_ASM in https://github.com/grpc/grpc/pull/16410, but got this error:

```
CMake Error at third_party/boringssl/CMakeLists.txt:182 (add_compile_options):
  Error evaluating generator expression:
    $<COMPILE_LANGUAGE:CXX>
  $<COMPILE_LANGUAGE:...> may not be used with Visual Studio generators.
```

**Comment by davidben** [comment]
> making those pre-build binaries CPU-architecture dependent could break users.

Binaries are always specific to CPU-architecture if you mean x86 vs x86_64, etc. If you're worried about being AVX2-specific or whatever, the assembly doesn't do that. Everything is conditioned at runtime on CPU feature detection as needed. Remember that Chrome ships binaries too and ships the assembly just fine.

> `$<COMPILE_LANGUAGE:...> may not be used with Visual Studio generators.`

That's hilarious, since that's used as part of the workaround for the CMake bug. Looking around, it seems CMake only added support for this in 3.11, so you may need to update your CMake to 3.11 or later. (Or condition `OPENSSL_NO_ASM` on the CMake version.)
https://gitlab.kitware.com/cmake/cmake/merge_requests/1657

Support for NASM in the Visual Studio generator itself was added in 3.9.
https://gitlab.kitware.com/cmake/cmake/merge_requests/428

Given my own CMake patch will be in 3.13, that's only a small window of releases the workaround is useful for. Perhaps we should remove it after 3.13 is released...
https://gitlab.kitware.com/cmake/cmake/merge_requests/2179

> do you want to backport the fix to the chromium-stable branch?

In general, the answer is no. The chromium-stable branch includes is what's shipped to Chrome. Any backported change must meet Chrome's [merge requirements](https://www.chromium.org/developers/the-zen-of-merge-requests), and this does not fix a problem for Chrome.

That said, this file is also unused by Chrome, so there's no meaningful risk to Chrome. There is, however, risk to other projects also consuming chromium-stable so this is, in some sense, worse since Chrome isn't testing the file for you. And, as you note, apparently this change broke the MSVC generator on older versions of CMake, so this probably shouldn't be merged...

Consumers should be updating from master if they want bugfixes and new features at a timely rate. The chromium-stable branch is for people who want a 12 week delay on all changes. Indeed had you not tried out the master branch, you wouldn't have noticed until that change made it to chromium-stable with Chrome 69 and fixing it would need to backport a revert. (Something we'd probably be willing to do but, again, backports are conditional on the file in question not being used by Chrome.)

I believe when this problem previously came up, your team said you would set up a process to continually test gRPC against BoringSSL master. Did this not happen? Paradoxically, consumers using chromium-stable tend to have *more* build failures due to this lack of feedback.

**Comment by jtattermusch** [comment]
@davidben  thank a lot for the insights!
>Binaries are always specific to CPU-architecture if you mean x86 vs x86_64, etc. If you're worried about being AVX2-specific or whatever, the assembly doesn't do that. Everything is conditioned at runtime on CPU feature detection as needed. Remember that Chrome ships binaries too and ships the assembly just fine.

I didn't know that. So just to confirm, you're saying that the only architectures recognized by boringssl at compile time are `x86`, `x86_64`, `aarch64`, `arm` and `ppc64le` (based on 
https://github.com/google/boringssl/blob/4ac9405ebae49b885b27f33f81b76e9e4584f386/CMakeLists.txt#L321) and everything more specific (such as special instruction sets available) is autodetected at runtime?
If that's the case, that would mean we could improve the performance (for secure channels) for some of our languages (e.g. for C# that ships a pre-compiled binaries in nuget packages) quite a bit without being worried about breaking users that have CPUs with different instruction sets - we're already differentiating between x86 and x86_64 but we thought a finer granularity would be needed if assembly optimization were to be enabled.
For reference, the umbrella issue for secure channel slowness with OPENSSL_NO_ASM=ON is https://github.com/grpc/grpc/issues/9440. Needless to say, we'd love to resolve that issue, as the potential performance gains are huge.

**Comment by davidben** [comment]
> So just to confirm, you're saying that the only architectures recognized by boringssl at compile time are x86, x86_64, aarch64, arm and ppc64le

Yes, with the following exceptions:

* Of course, if you pass GCC flags that tell it to assume, e.g., AVX support then the compiler may use those instructions in random loops and your binary will depend on AVX. But that's not BoringSSL.

* x86 assembly does *not* runtime-detect SSE2. That one is assumed. (Because OpenSSL assembly doesn't runtime-detect it.) But that's not much of an assumption; Chrome binaries also require it, and it's common to tell C compilers to freely emit SSE2 instructions. SSE2 is almost 20 years old now.

* ARM is weird because the feature checking logic is OS-specific rather than there being a common instruction on the architecture. (The relevant instruction in ARM is kernel-only and the kernel is supposed to expose it to userspace how it likes.) On Linux/Android, you check `getauxval` plus some mess of `/proc` workarounds for older Androids. We support that. On iOS and funnier embedded platforms, the expectation is that you produce multiple versions of your binary with specific features burned in. When configured that we, [we do that too](https://boringssl.googlesource.com/boringssl/+/HEAD/BUILDING.md#embedded-arm).

> Needless to say, we'd love to resolve that issue, as the potential performance gains are huge.

Yup. The expectation is definitely that assembly is enabled. It's there for a reason. I certainly hope worries about CPU-specific binaries wasn't what was holding this back!  Please reach out to us when you find yourself needing to disable something like this. Most of gRPC's problems around shipping binaries or reducing build dependencies have been solved by other teams already (Chrome, in particular, has a lot of experience here), so we can point you to things that have worked.

**Comment by davidben** [comment]
> Of course, if you pass GCC flags that tell it to assume, e.g., AVX support then the compiler may use those instructions in random loops and your binary will depend on AVX. But that's not BoringSSL.

I guess I should add: we don't do this much now (outside of a couple corners of ARM), but we may in the future short-circuit some runtime checks when the corresponding compiler flag is passed in. (Those flags set corresponding preprocessor symbols.) This would mostly be so consumers targeting only newer CPUs don't need to carry around code that only runs on older CPUs.

But this would only kick in if you explicitly tell the compiler to do such targeting, so you would already be assuming the feature by allowing the compiler to auto-vectorize loops, etc. So that should be a free optimization.

**Comment by jtattermusch** [comment]
let's keep this open as a reminder to change the behavior in the future, once we're fully migrated to VS2017

**Comment by davidben** [comment]
I think the requirement is a newer CMake version, not a newer Visual Studio.

**Comment by jtattermusch** [comment]
@zackgalbreath we could do this but only if it doesn't require bumping required cmake version too high (it seems that it might).

**Comment by davidben** [comment]
You can probably query the cmake version within cmake, right? Another option is to do it opportunistically based on the cmake version and maybe print a warning to update it for better security and performance.

Although since it's Windows, where folks are installing cmake directly anyway, or from some package manager hopefully more up-to-date than Debian, I would speculate that recent cmake versions on Windows are less of a hassle.

**Comment by jtattermusch** [comment]
Fixed by #21427.
